File: 500.键盘行.py
#
# @lc app=leetcode.cn id=500 lang=python
#
# [500] 键盘行
#
# 解法1(T64% S33%): 很朴素，就是每个单词的每个字母依次判断在哪行里，如果同一个单词的所有字母都在同一行里就加到结果中（通过flag和break控制）
# 
# 解法2(T98% S60%)：python库函数，通过`strip()`把word中所有第x行字符删掉，如果能删干净那证明这个word的所有字符都在同一行里

# @lc code=start
class Solution(object):
    def findWords(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        res = []
        for word in words:
            lword = word.lower()
            if not lword.strip("qwertyuiop") or not lword.strip("asdfghjkl") or not lword.strip("zxcvbnm"):
                res.append(word)
        return res


    def otherSolution(self, words):
        keyboard = ["qwertyuiop", "asdfghjkl", "zxcvbnm"]
        res = []
        for word in words:
            flag = True
            our = -1
            for l in word:
                for row in keyboard:
                    if l.lower() in row:
                        if our != -1 and row != our:
                            flag = False
                            break
                        our = row
                if not flag:
                    break

            if flag:
                res.append(word)

        return res
# @lc code=end

File: javascript/500.键盘行.js
/*
 * @lc app=leetcode.cn id=500 lang=javascript
 *
 * [500] 键盘行
 * 
 * 题不难，但代码还是很绕的，想好一种解决办法就去实现，不要在一套代码里出现两种解决思路，会混乱的
 */

// @lc code=start
/**
 * @param {string[]} words
 * @return {string[]}
 */
var findWords = function(words) {
    var sets = [new Set("qwertyuiop".split("")), new Set("asdfghjkl".split("")), new Set("zxcvbnm".split(""))];

    var res = [];
    for (var word of words) {
        var i=0;
        for (; i<sets.length; i++) {
            if (sets[i].has(word[0].toLowerCase())) {
                break;
            }
        }
        var flag = true;
        for (var j=1; j<word.length; j++) {
            if (!sets[i].has(word[j].toLowerCase())) {
                flag = false;
                break;
            }
        }

        if (flag) {
            res.push(word);
        }
    }
    return res;
};
// @lc code=end

